### PR TITLE
introduce the stm32h7rs serie 

### PR DESCRIPTION
**Pull request by FRASTM**
This PR is introducing the new stm32h7RS serie from STMicroelectronics 
https://www.st.com/en/microcontrollers-microprocessors/stm32h7r7-7s7.html

Based on the stm32H7 with 64KB of internal Flash memory and  Arm Cortex-M7 MCU

Requires https://github.com/zephyrproject-rtos/hal_stm32/pull/200

The stm32h7s78_dk is the disco kit with a external   1‑Gbit Octo‑SPI NOR flash memory (MX66uw1g45)

### CODE REVIEW

**Comment by FRASTM** [comment]
rebase including the hal_stm32 05ee82ff431deefc32322cccf112b9b6aff4aec6

File: soc/st/stm32/soc.yml [hunk 1]
```diff
@@ -101,6 +101,12 @@ family:
     - name: stm32h562xx
     - name: stm32h563xx
     - name: stm32h573xx
+  - name: stm32h7rsx
+    socs:
+    - name: stm32h7r3xx
+    - name: stm32h7r7xx
+    - name: stm32h7s3xx
+    - name: stm32h7s7xx
```
Comment (erwango) [inline soc/st/stm32/soc.yml:109]: Could it be moved to initial commit ?

File: include/zephyr/dt-bindings/reset/stm32h7_reset.h [hunk 2]
```diff
@@ -12,12 +13,18 @@
...
 #define STM32_RESET_BUS_AHB1  0x80
 #define STM32_RESET_BUS_AHB2  0x84
+#if defined(CONFIG_SOC_SERIES_STM32H7RSX)
+#define STM32_RESET_BUS_AHB3  0xA4
+#define STM32_RESET_BUS_AHB5  0x7C
+#define STM32_RESET_BUS_APB5  0x8C
+#else
 #define STM32_RESET_BUS_AHB3  0x7C
+#define STM32_RESET_BUS_APB3  0x8C
+#endif /* CONFIG_SOC_SERIES_STM32H7RSX */
```
Comment (erwango) [inline include/zephyr/dt-bindings/reset/stm32h7_reset.h:23]: KConfig symbols don't work in dts bindings (dts is parsed before Kconfig), a dedicated file is required

File: dts/arm/st/h7rs/stm32h7rs.dtsi [hunk 3]
```diff
@@ -0,0 +1,831 @@
...
+			gpiop: gpio@58023c00 {
+				compatible = "st,stm32-gpio";
+				gpio-controller;
+				#gpio-cells = <2>;
+				reg = <0x58023c00 0x400>;
+				clocks = <&rcc STM32_CLOCK_BUS_AHB4 0x00008000>;
+			};
+		};
+
+		iwdg: iwdg1: watchdog@58004800 {
```
Comment (erwango) [inline dts/arm/st/h7rs/stm32h7rs.dtsi:309]: Please limit dtsi file to basic nodes for now.
Then we'll add new hw block nodes by sequence with matching board configuration and test when possible

File: drivers/gpio/gpio_stm32.c [hunk 4]
```diff
@@ -326,12 +336,21 @@ static void gpio_stm32_set_exti_source(int port, int pin)
...
 #elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32g0_exti)
 	LL_EXTI_SetEXTISource(port, line);
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7rs_exti)
+	/* 'port' is of LL_SBS_EXTI_PORTx, 'line' is the pin value [0..15] */
+	/* Using HAL formula when setting the gpio port during HAL_EXTI_SetConfigLine */
+	uint32_t regval = SBS->EXTICR[line >> 2U];
+
+	regval &= ~(SBS_EXTICR1_PC_EXTI0 << (SBS_EXTICR1_PC_EXTI1_Pos * (line & 0x03u)));
+	regval |= (port << (SBS_EXTICR1_PC_EXTI1_Pos * (line & 0x03u)));
+	SBS->EXTICR[line >> 2u] = regval;
```
Comment (erwango) [inline drivers/gpio/gpio_stm32.c:346]: Why not using `LL_SBS_SetEXTISource` ?

File: drivers/gpio/gpio_stm32.c [hunk 5]
```diff
@@ -292,18 +292,28 @@ static int gpio_stm32_clock_request(const struct device *dev, bool on)
...
 	return ((pin % 4 * 4) << 16) | (pin / 4);
 #elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32g0_exti)
 	return ((pin & 0x3) << (16 + 3)) | (pin >> 2);
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7rs_exti)
+	/*
+	 * No need to translate pin to a LL_SBS_EXTI_LINEn value
+	 * as the gpio_stm32_set_exti_source does not call LL_SBS_SetEXTISource
+	 * but uses HAL formula : keep pin value [0..15]
+	 */
+	return (uint32_t)pin;
```
Comment (erwango) [inline drivers/gpio/gpio_stm32.c:309]: This operation is not in line anymore with the function name.
If changes have to be done in the driver implementation, it should be applied to the other series as well.

File: drivers/clock_control/clock_stm32_ll_h7.c [hunk 6]
```diff
@@ -843,12 +1031,13 @@ int stm32_clock_control_init(const struct device *dev)
 	defined(CONFIG_SOC_STM32H7B0XX) || defined(CONFIG_SOC_STM32H7B0XXQ) || \
 	defined(CONFIG_SOC_STM32H7B3XX) || defined(CONFIG_SOC_STM32H7B3XXQ)
 	LL_AHB2_GRP1_EnableClock(LL_AHB2_GRP1_PERIPH_HSEM);
-#else
-	LL_AHB4_GRP1_EnableClock(LL_AHB4_GRP1_PERIPH_HSEM);
-#endif
```
Comment (erwango) [inline drivers/clock_control/clock_stm32_ll_h7.c:848]: Changes impacting H7 series should be made at least in a dedicated commit, ideally in a dedicated PR

File: drivers/clock_control/clock_stm32_ll_h7.c [hunk 7]
```diff
@@ -874,19 +1063,24 @@ int stm32_clock_control_init(const struct device *dev)
 	/* AHB/AXI/HCLK clock is SYSCLK / HPRE */
 	/* If freq increases, set flash latency before any clock setting */
 	if (new_hclk_freq > old_hclk_freq) {
-		LL_SetFlashLatency(new_hclk_freq);
+		LL_FLASH_SetLatency(LL_FLASH_LATENCY_6);
```
Comment (erwango) [inline drivers/clock_control/clock_stm32_ll_h7.c:1066]: Changes impacting H7 series should be made at least in a dedicated commit, ideally in a dedicated PR

File: drivers/clock_control/clock_stm32_ll_h7.c [hunk 8]
```diff
@@ -874,19 +1063,24 @@ int stm32_clock_control_init(const struct device *dev)
...
+	LL_RCC_SetAPB2Prescaler(apb2_prescaler(STM32_D2PPRE));
+	LL_RCC_SetAPB4Prescaler(apb4_prescaler(STM32_D4PPRE));
+	LL_RCC_SetAPB5Prescaler(apb5_prescaler(STM32_D5PPRE));
+#else
 	LL_RCC_SetAPB1Prescaler(apb1_prescaler(STM32_D2PPRE1));
 	LL_RCC_SetAPB2Prescaler(apb2_prescaler(STM32_D2PPRE2));
-	LL_RCC_SetAPB3Prescaler(apb3_prescaler(STM32_D1PPRE));
 	LL_RCC_SetAPB4Prescaler(apb4_prescaler(STM32_D3PPRE));
-
+#endif
```
Comment (erwango) [inline drivers/clock_control/clock_stm32_ll_h7.c:1083]: Changes impacting H7 series should be made at least in a dedicated commit, ideally in a dedicated PR

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/bindings/clock/st,stm32h7rs-pll-clock.yaml [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+  It can be used to describe 3 different PLLs: PLL1 (Main PLL), PLL2 and PLL3.
+  Only PLL1 and PLL3 are supported for now.
+
+  These PLLs could take one of clk_hse, clk_hsi or clk_csi as input clock, with
+  an input frequency from 1 to 16 MHz. PLLM factor is used to set the input
+  clock in this acceptable range.
+
+  Each PLL can have up to 5 output clocks and for each output clock, the
+  frequency can be computed with the following formulae:
```
[the remark this answers is not in the document]
Comment (FRASTM) [reply dts/bindings/clock/st,stm32h7rs-pll-clock.yaml:15]: yes, up to 5. Adding the comment for PLL 2:
`f(PLL_Tx) = f(VCOx clock) / PLLTx   -> pllx_t_ck (only for PLL2)`

File: dts/bindings/clock/st,stm32h7rs-pll-clock.yaml [hunk 10]
```diff
@@ -0,0 +1,41 @@
+# Copyright (c) 2024 STMicroelectronics
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  PLL node binding for STM32H7RS devices
+
+  It can be used to describe 3 different PLLs: PLL1 (Main PLL), PLL2 and PLL3.
+  Only PLL1 and PLL3 are supported for now.
```
[the remark this answers is not in the document]
Comment (FRASTM) [reply dts/bindings/clock/st,stm32h7rs-pll-clock.yaml:8]: removed

File: dts/bindings/interrupt-controller/st,stm32h7rs-exti.yaml [hunk 11]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2024, STMicroelectronics
+
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+        STM32 controller
+        This compatible stands for the stm32H7RS interrupt-controller block
+        with two dedicated Rising and Falling interrupt pending registers
+        when the port is LL_SBS_EXTI_PORTx value and line is LL_SBS_EXTI_LINEn
```
[the remark this answers is not in the document]
Comment (FRASTM) [reply dts/bindings/interrupt-controller/st,stm32h7rs-exti.yaml:9]: removing the reference to LL driver

File: include/zephyr/dt-bindings/reset/stm32h7_reset.h (hunk 2 above)
Comment (FRASTM) [reply include/zephyr/dt-bindings/reset/stm32h7_reset.h:23]: new _include/zephyr/dt-bindings/reset/stm32h7rs_reset.h_ created instead

File: dts/arm/st/h7rs/stm32h7rs.dtsi (hunk 3 above)
Comment (FRASTM) [reply dts/arm/st/h7rs/stm32h7rs.dtsi:309]: removing adc, wdg, rng, dma nodes from this first PR

File: drivers/gpio/gpio_stm32.c (hunk 4 above)
Comment (FRASTM) [reply drivers/gpio/gpio_stm32.c:346]: change formula to use LL_SBS_SetEXTISource

File: drivers/gpio/gpio_stm32.c (hunk 5 above)
Comment (FRASTM) [reply drivers/gpio/gpio_stm32.c:309]: changed due to   LL_SBS_SetEXTISource

File: soc/st/stm32/soc.yml (hunk 1 above)
Comment (FRASTM) [reply soc/st/stm32/soc.yml:109]: moved in the commit d1715d0cafbd9b9dd7ac1194e7a5c12a421a6ac0

File: drivers/clock_control/clock_stm32_ll_h7.c (hunk 6 above)
Comment (FRASTM) [reply drivers/clock_control/clock_stm32_ll_h7.c]: remove change related to non-stm32h7rs

File: drivers/clock_control/clock_stm32_ll_h7.c (hunk 7 above)
Comment (FRASTM) [reply drivers/clock_control/clock_stm32_ll_h7.c]: remove change related to non-stm32h7rs

File: drivers/clock_control/clock_stm32_ll_h7.c (hunk 8 above)
Comment (FRASTM) [reply drivers/clock_control/clock_stm32_ll_h7.c]: remove change related to non-stm32h7rs

File: drivers/gpio/gpio_stm32.c [hunk 12]
```diff
@@ -803,3 +825,16 @@ GPIO_DEVICE_INIT_STM32(j, J);
 #if DT_NODE_HAS_STATUS(DT_NODELABEL(gpiok), okay)
 GPIO_DEVICE_INIT_STM32(k, K);
 #endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(gpiok), okay) */
+
```
Comment (erwango) [inline drivers/gpio/gpio_stm32.c:828]: Remove blank line

File: drivers/gpio/gpio_stm32.c [hunk 13]
```diff
@@ -341,7 +351,11 @@ static int gpio_stm32_get_exti_source(int pin)
 	port = LL_GPIO_AF_GetEXTISource(line);
 #elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32g0_exti)
 	port = LL_EXTI_GetEXTISource(line);
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7rs_exti)
+	/* 'port' is the LL_SBS_EXTI_PORTx, 'line' is LL_SBS_EXTI_LINEn */
```
Comment (erwango) [inline drivers/gpio/gpio_stm32.c:355]: Comment not update to date

File: drivers/gpio/gpio_stm32.c [hunk 14]
```diff
@@ -717,6 +735,10 @@ static int gpio_stm32_init(const struct device *dev)
 	/* Cf: L4/L5 RM, Chapter "Independent I/O supply rail" */
 	LL_PWR_EnableVddIO2();
 	z_stm32_hsem_unlock(CFG_HW_RCC_SEMID);
+#elif defined(CONFIG_SOC_SERIES_STM32H7RSX)
+	/* TBC : stm32Cube does */
+	LL_PWR_EnableXSPIM1();
+	LL_PWR_EnableUSBVoltageDetector();
```
Comment (erwango) [inline drivers/gpio/gpio_stm32.c:741]: Does it really relate to exti ?
Should it be in soc ? xspi ?

File: west.yml [hunk 15]
```diff
@@ -229,7 +229,7 @@ manifest:
       groups:
         - hal
     - name: hal_stm32
-      revision: 366c5c909fbcecf94f32411887132024f61b0bcd
+      revision: 077e943b60f95fab515ddfcbb41210b863b76951
```
Comment (erwango) [inline west.yml:232]: This commit should be removed, we already point on 54724a815d23b7f518fa88eb5fb7e991109e191b which integrate h7rs hal

**Review by erwango — CHANGES_REQUESTED** [review]
I need more time to review the clock control part

File: west.yml (hunk 15 above)
Comment (FRASTM) [reply west.yml]: refer to https://github.com/zephyrproject-rtos/zephyr/pull/73643

**Comment by FRASTM** [comment]
CI failure due to missing hal_stm32 missing the stm32h7rs serie

File: drivers/gpio/gpio_stm32.c (hunk 14 above)
Comment (FRASTM) [reply drivers/gpio/gpio_stm32.c]: useless, removing

File: drivers/gpio/gpio_stm32.c (hunk 13 above)
Comment (FRASTM) [reply drivers/gpio/gpio_stm32.c]: changed

File: include/zephyr/drivers/clock_control/stm32_clock_control.h [hunk 16]
```diff
@@ -86,13 +89,22 @@
...
 
-/** STM2H7 specifics RCC dividers */
+/** STM2H7RS specifics RCC dividers */
+#if defined(CONFIG_SOC_SERIES_STM32H7RSX)
+#define STM32_D1CPRE	DT_PROP(DT_NODELABEL(rcc), dcpre)
+#define STM32_HPRE	DT_PROP(DT_NODELABEL(rcc), hpre)
+#define STM32_D1PPRE	DT_PROP(DT_NODELABEL(rcc), ppre1)
+#define STM32_D2PPRE	DT_PROP(DT_NODELABEL(rcc), ppre2)
+#define STM32_D4PPRE	DT_PROP(DT_NODELABEL(rcc), ppre4)
+#define STM32_D5PPRE	DT_PROP(DT_NODELABEL(rcc), ppre5)
```
[the remark this answers is not in the document]
Comment (erwango) [reply include/zephyr/drivers/clock_control/stm32_clock_control.h:99]: This comment still stands

**Review by erwango — DISMISSED** [review]

**Comment by erwango** [comment]
Please remove the last commit ....

**Comment by FRASTM** [comment]
> Please remove the last commit ....

removed

**Review by erwango — APPROVED** [review]

**Review by ajarmouni-st — APPROVED** [review]

**Review by gautierg-st — APPROVED** [review]

**Comment by erwango** [comment]
Taking assignment as @galak is not much active these days.
